### PR TITLE
Functionality for grouping contacts

### PR DESCRIPTION
**Pull request by AlyaNovikova**
I accidentally merged with the previous branch, so there are old commits :(

### CODE REVIEW

File: alyanovikova/Contacts/Sources/Contacts/Contacts.swift [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+    contacts[newContact.id] = newContact
+
+    try writeToFile()
+  }
+
+  public func removeContact(id: Contact.Id) throws {
+    contacts.removeValue(forKey: id)
+
+    try writeToFile()
+  }
```
Comment (gartemiy) [inline alyanovikova/Contacts/Sources/Contacts/Contacts.swift:111]: please consider using `Contact` for `removeContact(_ contact: Contact) throws` and similar for groups.

File: alyanovikova/Contacts/Sources/Contacts/Contacts.swift [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+  public func listContacts(where predicate: (Contact) -> Bool) -> [Contact] {
+    return contacts.values.filter { predicate($0) }
+  }
+
+  // Functions with Group
+  public func addGroup(title: String, members: [Contact.Id] = []) throws -> Group {
```
Comment (gartemiy) [inline alyanovikova/Contacts/Sources/Contacts/Contacts.swift:118]:
```suggestion
  public func addGroup(title: String, contacts: [Contact] = []) throws -> Group {
```

File: alyanovikova/Contacts/Sources/Contacts/Contacts.swift [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+      throw ContactError.nonexistentIdOfContact
+    }
+
+    existGroup.remove(contactId: contact.id)
+    groups[group.id] = existGroup
+
+    try writeToFile()
+  }
+
+  public func members(of group: Group) throws -> [Contact] {
```
Comment (gartemiy) [inline alyanovikova/Contacts/Sources/Contacts/Contacts.swift:167]:
```suggestion
  public func contacts(of group: Group) throws -> [Contact] {
```

File: alyanovikova/Contacts/Sources/Contacts/Contacts.swift [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    self.phone = phone
+  }
+}
+
+public struct Group: Codable {
+  public typealias Id = UUID
+
+  public let id: Id
+  public var title: String
+  public var members: [Contact.Id]
```
Comment (gartemiy) [inline alyanovikova/Contacts/Sources/Contacts/Contacts.swift:26]:
```suggestion
  public var contactIds: [Contact.Id]
```

Comment (gartemiy) [inline alyanovikova/Contacts/Sources/Contacts/Contacts.swift:26]: Do you allow publicly to change the `title` of the `Group` or `members`?

**Comment by AlyaNovikova** [comment]
done!

File: alyanovikova/Contacts/Sources/Contacts/Contacts.swift [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+    contacts[newContact.id] = newContact
+
+    try writeToFile()
+  }
+
+  public func removeContact(id: Contact.Id) throws {
+    contacts.removeValue(forKey: id)
+
+    try writeToFile()
+  }
```
Comment (gartemiy) [inline alyanovikova/Contacts/Sources/Contacts/Contacts.swift]: Maybe remove this one as obsolete.

File: alyanovikova/Contacts/Tests/ContactsTests/XCTestManifests.swift [hunk 6]
```diff
@@ -0,0 +1 @@
+import XCTest
```
Comment (gartemiy) [inline alyanovikova/Contacts/Tests/ContactsTests/XCTestManifests.swift:1]: How about removing this file completely?

File: alyanovikova/Contacts/Tests/ContactsTests/ContactsTests.swift [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+      try XCTAssertEqual(book.contacts(of: friends), [man1])
+
+      try book.removeContact(id: man1.id)
+
+      try XCTAssertEqual(book.contacts(of: all), [man2])
+      try XCTAssertEqual(book.contacts(of: friends), [])
+    }
+  }
+
+  func testChangeTitle() throws {
```
Comment (gartemiy) [inline alyanovikova/Contacts/Tests/ContactsTests/ContactsTests.swift:344]: I think the name of the test is not descriptive enough. The same for `testContacts`

File: alyanovikova/Contacts/Tests/ContactsTests/ContactsTests.swift [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+        throw ContactsTests.ContactError.noGroup
+      }
+
+      guard let friends = book.groups[group1.id] else {
+        throw ContactsTests.ContactError.nonexistentIdOfGroup
+      }
+
+      guard let all = book.groups[group2.id] else {
+        throw ContactsTests.ContactError.nonexistentIdOfGroup
+      }
```
Comment (gartemiy) [inline alyanovikova/Contacts/Tests/ContactsTests/ContactsTests.swift:230]: Hm, there is no need to throw exception from API here, you can finish the test with XCTFail(...).

**Pull request merged**